### PR TITLE
Parser fixes

### PR DESCRIPTION
**Pull request by harendra-kumar**

### CODE REVIEW

File: src/Streamly/Internal/Data/Stream/StreamD/Eliminate.hs [hunk 1]
```diff
@@ -206,20 +217,13 @@ parse_ (PRD.Parser pstep initial extract) (Stream step state) = do
                 let (src0, buf1) = splitAt n (x:getList buf)
                     src  = Prelude.reverse src0 ++ xs
                 gobuf SPEC s (List buf1) (List src) pst1
-            PR.Done _ b -> return (b, Stream step_rest (Left (s, buf)))
+            PR.Done n b -> do
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD/Eliminate.hs:220]: Have a `Done 0` case here as well?

File: src/Streamly/Internal/Data/Parser/ParserD/Type.hs [hunk 2]
```diff
@@ -424,7 +424,7 @@ noErrorUnsafeSplitWith func (Parser stepL initialL extractL)
             -- Assume that the first parser can never fail, therefore we do not
             -- need to keep the input for backtracking.
             Partial n s -> return $ Partial n (SeqParseL s)
-            Continue n s -> return $ Partial n (SeqParseL s)
+            Continue n s -> return $ Continue n (SeqParseL s)
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Parser/ParserD/Type.hs:427]: This should be `Partial`. `noErrorUnsafeSplitWith` isf or parsers that don't fail.

File: src/Streamly/Internal/Data/Parser/ParserD/Type.hs [hunk 3]
```diff
@@ -438,7 +438,7 @@ noErrorUnsafeSplitWith func (Parser stepL initialL extractL)
         r <- stepR st a
         return $ case r of
             Partial n s -> Partial n (SeqParseR f s)
-            Continue n s -> Partial n (SeqParseR f s)
+            Continue n s -> Continue n (SeqParseR f s)
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Parser/ParserD/Type.hs:441]: This should be `Partial`. `noErrorUnsafeSplitWith` isf or parsers that don't fail.

File: src/Streamly/Internal/Data/Parser/ParserD/Type.hs (hunk 2 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Parser/ParserD/Type.hs:427]: Continue does not mean failure, its only backtracking. The inner parser wants to backtrack but you are dropping the backtrack buffer and continuing from the current position that would make the behavior of the parser incorrect.

File: src/Streamly/Internal/Data/Stream/StreamD/Eliminate.hs (hunk 1 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Stream/StreamD/Eliminate.hs:220]: Yes, that would avoid an unnecessary `append` which is detrimental to perf.

**Review by adithyaov — COMMENTED** [review]
You can merge it after adding a `Done 0` case.

File: src/Streamly/Internal/Data/Stream/StreamD/Eliminate.hs (hunk 1 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Stream/StreamD/Eliminate.hs:220]: the `Done 0` case won't be very useful here as the append would still be needed. And the rest of the code like `take` would anyway be simplified. So I am keeping it as is.

**Pull request merged**